core/domain/state_domain.py

```diff
@@ -916,8 +940,1085 @@ def get_all_outcomes(self) -> List[Outcome]:
+            'HasIntegerPartEqualTo',
+            'HasNoFractionalPart'
+        ]
+        lower_infinity = float('-inf')
+        upper_infinity = float('inf')
+        allow_non_zero_integ_part = (
+            self.customization_args[
+                'allowNonzeroIntegerPart'].value)
+        allow_imp_frac = (
+            self.customization_args[
+                'allowImproperFraction'].value)
+        require_simple_form = (
+            self.customization_args[
+                'requireSimplestForm'].value)
+        for ans_group_index, answer_group in enumerate(self.answer_groups):
+            for rule_spec_index, rule_spec in enumerate(
+                answer_group.rule_specs):
+                # Rule should not be duplicate.
+                if rule_spec.to_dict() in rule_spec_till_now:
+                    raise utils.ValidationError(
+                        f'The rule \'{rule_spec_index}\' of answer group '
+                        f'\'{ans_group_index}\' of FractionInput '
+                        f'interaction is already present.'
+                    )
+                else:
+                    rule_spec_till_now.append(rule_spec.to_dict())
+
+                if rule_spec.rule_type not in inputs_with_whole_nums:
+                    num = rule_spec.inputs['f']['numerator']
+                    den = rule_spec.inputs['f']['denominator']
+                    whole = rule_spec.inputs['f']['wholeNumber']
+
+                    # Denominator should be greater than 0.
+                    if whole == 0:
+                        if den <= 0:
+                            raise utils.ValidationError(
+                                f'The rule \'{rule_spec_index}\' of answer '
+                                f'group \'{ans_group_index}\' has '
+                                f'denominator less than or equal to zero '
+                                f'in FractionInput interaction.'
+                            )
+
+                    # Solution should be in simplest form if the `simplest form`
+                    # setting is turned on.
+                    if require_simple_form:
+                        if whole == 0:
+                            d = math.gcd(num, den)
+                            val_num = num // d
+                            val_den = den // d
+                            if val_num != num and val_den != den:
+                                raise utils.ValidationError(
+                                    f'The rule \'{rule_spec_index}\' of '
+                                    f'answer group \'{ans_group_index}\' do '
+                                    f'not have value in simple form '
+                                    f'in FractionInput interaction.'
+                                )
+
+                    # Solution should be in proper form if the `proper form`
+                    # setting is turned on.
+                    if not allow_imp_frac:
+                        if whole == 0:
```

Comment:
what if the proper form setting is turned on and the integer part is non-zero? What should we do? Is it an error?

Reply:
Yep! error should be raised. I have updated it